### PR TITLE
Implement websocket game server in JS

### PR DESCRIPTION
**Pull request by muteflamingo**
The game works, which is a plus to merging this, and also the linter was quiet. This one should be good to go.

### CODE REVIEW

**Comment by muteflamingo** [comment]
@maxbogue please merge

File: base.js [hunk 1]
```diff
@@ -24,4 +28,4 @@ function randomChoice(ls) {
   return ls[Math.floor(Math.random() * ls.length)];
 }
 
-module.exports = { max, min, randomChoice };
+module.exports = {gameID, max, min, randomChoice };
```
Comment (maxbogue) [inline base.js]: Space after the `{`

File: game.js [hunk 2]
```diff
@@ -139,4 +139,5 @@ class PlanetWars {
   }
 }
 
-module.exports = { PlanetWars };
+module.exports = { PlanetWars , AIS, MAPS};
```
Comment (maxbogue) [inline game.js]: Space before the `}` and no trailing line.

File: socketview.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    this.socket.emit('update', [planets, fleetJson]);
+  }
+
+  gameOver(winner, shipCounts) {
+    this.socket.emit('gameOver', winner, shipCounts);
+  }
+}
+
+module.exports = { WebsocketView };
+
```
Comment (maxbogue) [inline socketview.js]: Remove extra line.

File: socketview.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
+class WebsocketView {
+
```
Comment (maxbogue) [inline socketview.js]: Remove extra line.

File: package.json [hunk 5]
```diff
@@ -1,8 +1,8 @@
 {
-  "name": "planetwars",
-  "version": "1.0.0",
-  "description": "",
-  "main": "play.js",
+  "name": "pw_rewrite",
```
Comment (maxbogue) [inline package.json]: This shouldn't change from "planetwars".

File: package.json [hunk 6]
```diff
@@ -1,8 +1,8 @@
...
+  "version": "0.0.1",
+  "description": "planet wars but modern",
```
Comment (maxbogue) [inline package.json]: Maybe "implemented in node" instead of "but modern"?

File: socketview.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+  constructor(socket) {
+    this.socket = socket;
+  }
+
+  initialize(turnsPerSecond, planets, mapName, players) {
+    this.socket.emit('initialize', {turnsPerSecond, planets, mapName, players});
```
Comment (maxbogue) [inline socketview.js]: Spaces inside brackets.

File: socketview.js [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+  }
+
+  update(planets, fleets) {
+    let fleetJson = fleets.map((fleet) => fleet.forJson());
```
Comment (maxbogue) [inline socketview.js]: I typed `fleetJson` but it should really be `fleetsForJson`.

File: app.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+app.get('/game/:gameID', function(req, res) {
+  if (games[req.params.gameID]) {
+    res.render('game', {
+      gameID: req.params.gameID
+    });
+  } else {
+    res.redirect('/');
+  }
+});
+app.use('/static', express.static('static'));
```
Comment (maxbogue) [inline app.js]: `app.use` should be above the `app.get`/`app.post` and they all should have empty lines around them.

File: app.js [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+app.post('/create-game', function(req, res) {
+
+  let game_id = baseUtil.gameID();
```
Comment (maxbogue) [inline app.js]: This project uses camelCase not snake_case. Here and `turns_per_second` below.

File: app.js [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+  let p2 = req.body.p2;
+  let m = req.body.map;
+  if (m === 'Random') {
+    m = baseUtil.randomChoice(ldKeys(MAPS));
+  }
+  let turns_per_second = req.body.tps;
+  games[game_id] = new PlanetWars([p1, p2], m, turns_per_second);
+  res.redirect('/game/' + game_id);
+});
+
```
Comment (maxbogue) [inline app.js]: One empty line is enough, here and below.

File: package.json [hunk 12]
```diff
@@ -15,18 +15,17 @@
...
-  "bugs": {
-    "url": "https://github.com/maxbogue/planetwars/issues"
-  },
-  "homepage": "https://github.com/maxbogue/planetwars#readme",
+  "license": "ISC",
   "dependencies": {
-    "lodash": "^4.17.10"
+    "body-parser": "^1.18.3",
+    "express": "^4.16.3",
+    "express-generator": "^4.16.0",
```
Comment (maxbogue) [inline package.json]: Please remove express-generator, js-beautify, and uuid.

File: views/game.nunjs [hunk 13]
```diff
@@ -35,17 +35,20 @@
...
         function main() {
             var gameID = "{{ gameID }}";
             var socket = io.connect("/game");
             var canvas = document.getElementById("game");
             var ctx = canvas.getContext("2d");
-            socket.on("initialize", function(msg) {
-                var data = JSON.parse(msg);
+
+            socket.on("initialize", function(messagedata) {
+		console.log(messagedata);
```
Comment (maxbogue) [inline views/game.nunjs]: Remove console statements here and on line 51 and 61.

File: views/game.nunjs [hunk 14]
```diff
@@ -35,17 +35,20 @@
...
+                var data = messagedata
```
Comment (maxbogue) [inline views/game.nunjs]: Just call the arg `data` instead of `messagedata`, or `msg` for consistency with below.

File: static/test.html [hunk 15]
```diff
@@ -0,0 +1,16 @@
+
```
Comment (maxbogue) [inline static/test.html]: Extra file?

File: index.html [hunk 16]
```diff
@@ -0,0 +1,39 @@
+<!doctype html>
```
Comment (maxbogue) [inline index.html]: Extra file?

File: views/index.nunjs [hunk 17]
```diff
@@ -51,6 +51,8 @@
 </head>
 <body>
     <h1>Planet Wars</h1>
+	<h2>REWRITTEN</h2>
```
Comment (maxbogue) [inline views/index.nunjs]: Remove these lines.

File: base.js [hunk 18]
```diff
@@ -1,3 +1,7 @@
+const uuidv1 = require('uuid/v1');
+function gameID () {
+  return uuidv1();
```
[the remark this answers is not in the document]
Comment (muteflamingo) [reply base.js]: totes mcgotes

File: app.js (hunk 10 above)
Comment (maxbogue) [reply app.js]: Still applicable for `game_id`

File: app.js [hunk 19]
```diff
@@ -1,25 +1,32 @@
+//local imports, as requested
```
Comment (maxbogue) [inline app.js]: Libraries before local ones. You don't need comments for them.

File: app.js [hunk 20]
```diff
@@ -1,25 +1,32 @@
...
+const baseUtil = require('./base');
+const { PlanetWars, MAPS, AIS } = require('./game');
+const { WebsocketView } = require('./socketview');
+
+//require actual packages
 const bodyParser = require('body-parser');
+const _ = require('lodash');
 const express = require('express');
+
+
```
Comment (maxbogue) [inline app.js]: Extra blank line can be removed.

File: app.js [hunk 21]
```diff
@@ -1,25 +1,32 @@
...
-const baseUtil = require('./base');
-
-const {WebsocketView} = require('./socketview');
-
 const server = require('http').Server(app);
 const io = require('socket.io')(server);
-const { PlanetWars, MAPS, AIS } = require('./game');
-
 const games = {};
+//publish static files
```
Comment (maxbogue) [inline app.js]: Comments are good but there should 1) be a blank line above them all and 2) be a space after the `//`.

File: base.js [hunk 22]
```diff
@@ -1,6 +1,8 @@
-const uuidv1 = require('uuid/v1');
 function gameID () {
-  return uuidv1();
+  let digits = 'abcdef1234567890cw';
```
Comment (maxbogue) [inline base.js]: cw at the end looks like vim stuff snuck in.

File: base.js [hunk 23]
```diff
@@ -1,6 +1,8 @@
...
+  let id = randomChoice(digits) + randomChoice(digits) + randomChoice(digits);
+  id = id + randomChoice(digits) + randomChoice(digits);
```
Comment (maxbogue) [inline base.js]: You can just wrap a line and indent instead of having this be two lines.

File: package.json [hunk 24]
```diff
@@ -1,7 +1,7 @@
 {
-  "name": "pw_rewrite",
-  "version": "0.0.1",
-  "description": "planet wars but modern",
+  "name": "planetwars",
+  "version": "1.0.2",
```
Comment (maxbogue) [inline package.json]: Let's make it 0.1.0

**Comment by maxbogue** [comment]
Cool, thanks for coming back to this!

**Pull request merged**